### ISSUE TITLE
issues in changing "remove the redundant condition"

### ISSUE THREAD

**Issue by mychina75** [leohsuofnthu/Tensorflow-YOLACT#9]
Hi,
Noticed your changes in yolact_parser.py
deleted line 166:
        image = tf.cond(
            tf.equal(tf.shape(image)[-1], tf.constant(3)),
            true_fn=lambda: image,
            false_fn=lambda: tf.ones([image_width, image_height, 3])
        )
the change may cause issues in validation, when there are gray image.

have a nice day. :)

**Comment by leohsuofnthu** [comment]
Hi, Yes you are right
Thanks a lot ! will update ASAP

**Comment by leohsuofnthu** [comment]
Hi again, Hope ur training process goes well. A good news here, I manage to optimize the training speed with significant improvement now. U can train ur data quicker now :)

**Comment by mychina75** [comment]
sorry that I am new to TF2.
One more issue is currently the training always break, after some iters. maybe a epoch?

**Comment by leohsuofnthu** [comment]
it s okay, can you post it the error msg?

**Comment by mychina75** [comment]
no error. only exit(0)... It occurred 2 times. 
maybe there are some branch in training to make it exit?

**Comment by mychina75** [comment]
and about coco data. 
coco has 80 classes, but the catId ranges from 1 to 90. 
I am not sure in loss computing, if we should set the num_classes as 81.

there are always function to make the mapping, during compose the tfrecord. 
like: 
def _cat_id_to_real_id(readId):
  """Note coco has 80 classes, but the catId ranges from 1 to 90!"""
  cat_id_to_real_id = \
    {1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8, 9: 9, 10: 10, 11: 11, 13: 12, 14: 13, 15: 14, 16: 15, 17: 16,
     18: 17, 19: 18, 20: 19, 21: 20, 22: 21, 23: 22, 24: 23, 25: 24, 27: 25, 28: 26, 31: 27, 32: 28, 33: 29, 34: 30,
     35: 31, 36: 32, 37: 33, 38: 34, 39: 35, 40: 36, 41: 37, 42: 38, 43: 39, 44: 40, 46: 41, 47: 42, 48: 43, 49: 44,
     50: 45, 51: 46, 52: 47, 53: 48, 54: 49, 55: 50, 56: 51, 57: 52, 58: 53, 59: 54, 60: 55, 61: 56, 62: 57, 63: 58,
     64: 59, 65: 60, 67: 61, 70: 62, 72: 63, 73: 64, 74: 65, 75: 66, 76: 67, 77: 68, 78: 69, 79: 70, 80: 71, 81: 72,
     82: 73, 84: 74, 85: 75, 86: 76, 87: 77, 88: 78, 89: 79, 90: 80}
  return cat_id_to_real_id[readId]

**Comment by leohsuofnthu** [comment]
> no error. only exit(0)... It occurred 2 times.
> maybe there are some branch in training to make it exit?

I m not sure about this, maybe u could using tf.print to spot the bugs

**Comment by leohsuofnthu** [comment]
> and about coco data.
> coco has 80 classes, but the catId ranges from 1 to 90.
> I am not sure in loss computing, if we should set the num_classes as 81.
> 
> there are always function to make the mapping, during compose the tfrecord.
> like:
> def _cat_id_to_real_id(readId):
> """Note coco has 80 classes, but the catId ranges from 1 to 90!"""
> cat_id_to_real_id = 
> {1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8, 9: 9, 10: 10, 11: 11, 13: 12, 14: 13, 15: 14, 16: 15, 17: 16,
> 18: 17, 19: 18, 20: 19, 21: 20, 22: 21, 23: 22, 24: 23, 25: 24, 27: 25, 28: 26, 31: 27, 32: 28, 33: 29, 34: 30,
> 35: 31, 36: 32, 37: 33, 38: 34, 39: 35, 40: 36, 41: 37, 42: 38, 43: 39, 44: 40, 46: 41, 47: 42, 48: 43, 49: 44,
> 50: 45, 51: 46, 52: 47, 53: 48, 54: 49, 55: 50, 56: 51, 57: 52, 58: 53, 59: 54, 60: 55, 61: 56, 62: 57, 63: 58,
> 64: 59, 65: 60, 67: 61, 70: 62, 72: 63, 73: 64, 74: 65, 75: 66, 76: 67, 77: 68, 78: 69, 79: 70, 80: 71, 81: 72,
> 82: 73, 84: 74, 85: 75, 86: 76, 87: 77, 88: 78, 89: 79, 90: 80}
> return cat_id_to_real_id[readId]

Yes , I didn’t do this mapping during creating tf record, so If u doing on ur tfrecord , u can set the number of class to 80 when calculating the loss.

**Comment by mychina75** [comment]
ok. If the exit issue not occurred in your side, maybe my environment has some problem.
and yes, the training process looks faster than before. loss getting lower faster. 
great job~

**Comment by leohsuofnthu** [comment]
Thx !! Btw, when I creating then TFrecord, I filter out all the images that only hav crowd labels, cuz if there r images only with crowd annotation, after filter out by parser, there is no labels to train anymore. Don’t know if this relate to the situation u face now.

**Comment by mychina75** [comment]
yeah, maybe the tfrecord caused the issue....
I will check it. thank you~

**Comment by mychina75** [comment]
about training, after apply your optimize training code, training loss getting lower fast. 
but val loss getting worse. maybe getting overfitting? and before the optimization, value of val loss similiar with train loss... now the value is much bigger.

##########################
Saved checkpoint for step 250000: ./checkpoints/ckpt-26
Iteration 250001, **Train Loss: 8.95744800567627**, Loc Loss: 2.4022161960601807,  Conf Loss: 2.9272122383117676, Mask Loss: 2.732536554336548, Seg Loss: 0.8954824209213257
Iteration 250001, **Valid Loss: 16.83930015563965,** V Loc Loss: 4.657334327697754,  V Conf Loss: 6.5674357414245605, V Mask Loss: 4.232900619506836, Seg Loss: 1.381611704826355

Saved checkpoint for step 260000: ./checkpoints/ckpt-27
Iteration 260001, **Train Loss: 8.755365371704102**, Loc Loss: 2.3309450149536133,  Conf Loss: 2.878370761871338, Mask Loss: 2.6783549785614014, Seg Loss: 0.8677094578742981
Iteration 260001, **Valid Loss: 17.29536247253418**, V Loc Loss: 4.593933582305908,  V Conf Loss: 7.028153896331787, V Mask Loss: 4.272549152374268, Seg Loss: 1.4007411003112793

**Comment by leohsuofnthu** [comment]
Thanks, that s really important to know, I assume that something in data pipeline might have some problem. Will do the cross-check with pre-optimization code ASAP. Btw, what mahcine do u use, u train really quick.

**Comment by mychina75** [comment]
I am using GTX 1080Ti. 1 GPU card.

**Comment by leohsuofnthu** [comment]
Nice, I just found that I didnt normalize the images in valid set like in train set, and also re assure the randomness of dataloader. also tried small set of coco to see if there is overfitting, seems fine now. Thanks for point out :) Already update the code!

**Comment by mychina75** [comment]
great job~
